### ISSUE TITLE
:path header has to be an absolute path in push_promise frame

### ISSUE THREAD

**Issue by shekhei** [tatsuhiro-t/nghttp2#485]
I came across this when I pushed a :path header that is not an absolute path, I do not know if this is intentional, when I tried it with chrome, it was actually accepted, couldn't find this specific rule in the rfc either.

Below are two examples of failed request in nghttp
```
[  0.063] recv (stream_id=13) :method: GET
[  0.063] recv (stream_id=13) :path: ./global.js
[  0.063] recv (stream_id=13) :scheme: https
[  0.063] recv (stream_id=13) :authority: localhost:3000
[  0.063] recv (stream_id=13) accept: *.*
[  0.063] [INVALID; error=Violation in HTTP messaging rule] recv PUSH_PROMISE frame <length=33, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0, promised_stream_id=2)
[  0.063] recv (stream_id=13) :status: 200
[  0.063] recv HEADERS frame <length=1, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
```

```
[  0.064] recv (stream_id=13) :method: GET
[  0.064] recv (stream_id=13) :path: a/global.js
[  0.064] recv (stream_id=13) :scheme: https
[  0.064] recv (stream_id=13) :authority: localhost:3000
[  0.064] recv (stream_id=13) accept: *.*
[  0.064] [INVALID; error=Violation in HTTP messaging rule] recv PUSH_PROMISE frame <length=33, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0, promised_stream_id=2)
[  0.064] recv (stream_id=13) :status: 200
[  0.064] recv HEADERS frame <length=1, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
```

and below are two successful ones

```
[  0.062] recv (stream_id=13) :method: GET
[  0.062] recv (stream_id=13) :path: /./global.js
[  0.062] recv (stream_id=13) :scheme: https
[  0.062] recv (stream_id=13) :authority: localhost:3000
[  0.062] recv (stream_id=13) accept: *.*
[  0.062] recv PUSH_PROMISE frame <length=34, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0, promised_stream_id=2)
[  0.062] recv (stream_id=2) :status: 200
[  0.062] recv (stream_id=2) content-type: application/javascript
[  0.062] recv HEADERS frame <length=24, flags=0x24, stream_id=2>
          ; END_HEADERS | PRIORITY
          (padlen=0, dep_stream_id=0, weight=16, exclusive=0)
          ; First push response header
[  0.062] recv (stream_id=13) :status: 200
[  0.062] recv HEADERS frame <length=1, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0)
          ; First response header
```

```
[  0.062] recv (stream_id=13) :method: GET
[  0.062] recv (stream_id=13) :path: /global.js
[  0.062] recv (stream_id=13) :scheme: https
[  0.062] recv (stream_id=13) :authority: localhost:3000
[  0.062] recv (stream_id=13) accept: *.*
[  0.062] recv PUSH_PROMISE frame <length=33, flags=0x04, stream_id=13>
          ; END_HEADERS
          (padlen=0, promised_stream_id=2)
[  0.062] recv (stream_id=2) :status: 200
[  0.062] recv (stream_id=2) content-type: application/javascript
```

only the ones that had :path that's absolute worked.

**Comment by shekhei** [comment]
I believe i might have been mistaken, after some more testing, chrome seem to have sent an error too, closing it now, sorry for the mistake

**Comment by tatsuhiro-t** [comment]
For the record, RFC 7540 says about :path:

> The ":path" pseudo-header field includes the path and query parts
      of the target URI (the "path-absolute" production and optionally a
      '?' character followed by the "query" production (see Sections 3.3
      and 3.4 of [RFC3986]).  A request in asterisk form includes the
      value '*' for the ":path" pseudo-header field

And path-absolute production rule from RFC 3986 is:

```
path-absolute = "/" [ segment-nz *( "/" segment ) ]
```

We also allow asterisk-form here which is only used in OPTIONS method in practice.
To summarize, for push purposes, we probably don't use OPTIONS, so we need absolute path which starts "/".

**Comment by shekhei** [comment]
Ah i was trying to find the path absolute from the rfc, searched for a while, just couldnt find it for some odd reasons, thanks btw!
